lib/src/chunking.rs

```diff
@@ -372,82 +394,409 @@ impl Chunking {
-            .or_default()
-            .push(component);
+    let before_processing_pkgs_len = components.len();
+    if before_processing_pkgs_len == 0 {
+        return Vec::new();
     }
-    // Take all the non-tail (largest) components, and append them first
-    r.extend(components.into_iter().map(|v| vec![v]));
-    // Add the tail
-    r.extend(by_src.into_values());
-    // And order the new list
-    sort_packing(&mut r);
-    // It's possible that merging components gave us enough space; if so
-    // we're done!
-    if r.len() <= bins.get() as usize {
-        return r;
+    //Flatten out prior_build_metadata[i] to view all the packages in prior build as a single vec
+    //
+    //If the current rpm-ostree commit to be encapsulated is not the one in which packing structure changes, then
+    //  Compare flatten(prior_build_metadata[i]) to components to see if pkgs added, updated,
+    //  removed or kept same
+    //  if pkgs added, then add them to the last bin of prior[i][n]
+    //  if pkgs removed, then remove them from the prior[i]
+    //  iterate through prior[i] and make bins according to the name in nevra of pkgs and return
+    //  (no need of recomputing packaging structure)
+    //else if pkg structure to be changed || prior build not specified
+    //  Recompute optimal packaging strcuture (Compute partitions, place packages and optimize build)
+
+    if let Some(prior_build) = prior_build_metadata
+    /* && structure not be changed*/
+    {
+        println!("Keeping old package structure");
+        let mut curr_build: Vec<Vec<String>> = prior_build.clone();
+        //Packing only manaages RPMs not OStree commit
+        curr_build.remove(0);
+        let mut prev_pkgs: Vec<String> = Vec::new();
+        for bin in &curr_build {
+            for pkg in bin {
+                prev_pkgs.push(pkg.to_string());
+            }
+        }
+        prev_pkgs.retain(|name| !name.is_empty());
+        let curr_pkgs: Vec<String> = components
+            .iter()
+            .map(|pkg| pkg.meta.name.to_string())
+            .collect();
+        let prev_pkgs_set: HashSet<String> = HashSet::from_iter(prev_pkgs);
+        let curr_pkgs_set: HashSet<String> = HashSet::from_iter(curr_pkgs);
+        let added: HashSet<&String> = curr_pkgs_set.difference(&prev_pkgs_set).collect();
+        let removed: HashSet<&String> = prev_pkgs_set.difference(&curr_pkgs_set).collect();
+        let mut add_pkgs_v: Vec<String> = Vec::new();
+        for pkg in added {
+            add_pkgs_v.push(pkg.to_string());
+        }
+        let mut rem_pkgs_v: Vec<String> = Vec::new();
+        for pkg in removed {
+            rem_pkgs_v.push(pkg.to_string());
+        }
+        let curr_build_len = &curr_build.len();
+        curr_build[curr_build_len - 1].retain(|name| !name.is_empty());
+        curr_build[curr_build_len - 1].extend(add_pkgs_v);
+        for bin in curr_build.iter_mut() {
+            bin.retain(|pkg| !rem_pkgs_v.contains(pkg));
```

Comment:
`.contains()` on a `Vec` is O(N)...why are we not just looking up in `removed`?  Oh...are you hitting `&String` versus `&str` issues?